### PR TITLE
Update ver. 0.4 to the specification

### PR DESCRIPTION
**Pull request by swcurran**
Signed-off-by: Stephen Curran <swcurran@gmail.com>

### CODE REVIEW

**Comment by swcurran** [comment]
An update to the specification as per #106.

Note that the names of the items in the DID Log Entry object remain long form -- `versionId`, `versionTime`, `parameters`, `state` and `proof`.  There was some discussion of shortening them, but I plan to do that in a subsequent commit once we can have a discussion of the names to use.  Changing them from the current names to shorter versions will be easier from this point.

To be discussed at the 2024.09.26 did:tdw meeting -- [agenda here](https://hackmd.io/k4cIK9vQSlaeg2pdHE51IQ).

Feedback ahead of the meeting welcome -- anything unexpected in this?

File: spec/version.md [hunk 1]
```diff
@@ -2,6 +2,14 @@
 
 The following lists the substantive changes in each version of the specification.
 
+- Version 0.4
+  - Removes the use of JSON Patch from the specification. The full DIDDoc is included in each [[ref: DID log entry]].
+  - Changes the data format of the [[ref: DID log entries]] from an array to an object. The [[ref: DID Log]] remains in the [[ref: JSON Lines]] format.
```
Comment (martipos) [inline spec/version.md:7]:
```suggestion
  - Changes the data format of the [[ref: DID log entries]] from an array to an object. The [[ref: DID Log]] remains in the [[ref: JSON Lines]] format.
  - Change the data format of the Data Integrity Proof from an array to an object
```

**Comment by swcurran** [comment]
I've now updated the examples, per the implementation from the Python version.  In theory, this is now complete.

**Review by martipos — COMMENTED** [review]
Wording changes

- Removed my previous change that the Integrity Proof is a JSON Object, based on yesterdays discussion outcome (it will remain an array/set)
- replaced 'item'/'s' with 'property'/'ies' as we agreed to change the DID log entry data format  from array to JSON object

Signed-off-by: martipos <176692840+martipos@users.noreply.github.com>

File: spec/specification.md [hunk 2]
```diff
@@ -140,35 +141,38 @@ version of the DID is an update to the contents of the resolved [[ref: DIDDoc]]
 DID, and/or a change to the [[ref: parameters]] that control the generation and
 verification of the DID.
 
-Each entry is a JSON array consisting of the following 5 items.
+Each entry is a JSON object consisting of the following items.
 
-`[ versionId, versionTime, parameters, DIDDoc State, Data Integrity Proof ]`
+`{ "versionId": "", "versionTime": "", "parameters": {}, "state": {}, "proof" : {} }`
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:146]: Fixed

File: spec/specification.md [hunk 3]
```diff
@@ -140,35 +141,38 @@ version of the DID is an update to the contents of the resolved [[ref: DIDDoc]]
...
+   **MUST** be the time the DID will be retrieved by a [[ref: witness]] or resolver,
+   or before.
+3. The JSON object `parameters` contains the configurations/options set by the
+   [[ref: DID Controller]] to be used in the processing of current and future
+   [[ref: log entries]]. Permitted `parameters` are defined in the [`did:tdw`
+   DID Method Parameters](#didtdw-did-method-parameters) section of this
+   specification.
+4. The JSON object `state` contains the [[ref: DIDDoc]] for this version of the
+   DID.
+5. The JSON object `proof` is a [[ref: Data Integrity]] proof calculated across
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:167]: Fixed

File: spec/specification.md [hunk 4]
```diff
@@ -215,74 +220,71 @@ Creating a `did:tdw` DID is done by carrying out the following steps.
...
+   a literal `-`, followed by the `entryHash` value calculated in the previous
+   step.
+   5. **Generate the [[ref: Data Integrity]] proof** A [[ref: Data Integrity]]
+   proof on the preliminary JSON object as updated in the previous step **MUST**
+   be generated using an authorized key in the required `updateKeys` item in the
+   [[ref: parameters]] object.
+
+   If the [[ref: DID Controller]] has opted to use [[ref: witnesses]] for the
+   DID, the required approvals from the DID's [[ref: witnesses]] **MUST** be
+   collected and added to the [[ref: Data Integrity]] proof item. See the [DID
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:277]: Fixed

File: spec/specification.md [hunk 5]
```diff
@@ -677,13 +674,13 @@ To verify the [[ref: SCID]] of a `did:tdw` DID being resolved, the resolver
...
    in this process.
-2. Extract the `scid` item value from the [[ref: parameters]] in the [[ref: DID log entry]]. 
+2. Extract the `scid` item value from the [[ref: parameters]] in the [[ref: DID log entry]].
 3. Determine the hash algorithm used by the [[ref: DID Controller]] from the [[ref: multihash]] `scid` value.
    - The hash algorithm **MUST** be one listed in the
    [parameters](#didtdw-did-method-parameters) defined by the version of the
    `did:tdw` specification being used by the [[ref: DID Controller]] based on the
    `method` [[ref: parameters]] item.
-4. Remove the [[ref: data integrity]] proof item from the first [[ref: DID log entry]].
+4. Remove the [[ref: data integrity]] proof item from the [[ref: DID log entry]].
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:683]: Fixed

File: spec/specification.md [hunk 6]
```diff
@@ -766,7 +763,7 @@ Resolver **MUST** execute the following process:
 #### Authorized Keys
 
 Each entry in the [[ref: DID Log]] **MUST** include a [[ref: Data Integrity]]
-proof signed by a key **authorized** to control (create, update, deactivate) the
+`proof` item signed by a key **authorized** to control (create, update, deactivate) the
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:766]: Fixed

File: spec/version.md [hunk 7]
```diff
@@ -2,6 +2,15 @@
...
+  - Change the data format of the Data Integrity Proof from an array to an object
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/version.md:8]: Removed

File: spec/overview.md [hunk 8]
```diff
@@ -61,31 +61,29 @@ The following is a `tl;dr` summary of how `did:tdw` works:
    whitespace removed (per [[ref: JSON Lines]]). Each entry contains the
    information needed to derive a version of the [[ref: DIDDoc]] from its preceding
    version. The `did.jsonl` is also referred to as the [[ref: DID Log]].
-3. Each [[ref: DID log entry]] includes a JSON array of five items:
-    1. The `versionId` of the entry, a value that combines the version number
-       (starting at `1` and incrementing by one per version), a literal dash
+3. Each [[ref: DID log entry]] is a JSON object containing the following items:
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/overview.md:64]: Fixed

File: spec/specification.md [hunk 9]
```diff
@@ -140,35 +141,38 @@ version of the DID is an update to the contents of the resolved [[ref: DIDDoc]]
 DID, and/or a change to the [[ref: parameters]] that control the generation and
 verification of the DID.
 
-Each entry is a JSON array consisting of the following 5 items.
+Each entry is a JSON object consisting of the following items.
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:144]: Fixed

File: spec/specification.md [hunk 10]
```diff
@@ -215,74 +220,71 @@ Creating a `did:tdw` DID is done by carrying out the following steps.
...
-   The [DID Generation and Verification Parameters](#didtdw-did-method-parameters)
-   section of this specification defines the permitted [[ref: parameters]]. That
-   section defines what items **MUST** be included in this first [[ref: log entry]] for the DID.
-   Collect and use the specified [[ref: parameters]] in the remainder of the DID creation steps.
-
-   4. **Add the initial [[ref: DIDDoc]]** 
-   The fourth item in the input JSON array **MUST** be the JSON object `{"value": <diddoc> }`,
-   where `<diddoc>` is the initial [[ref: DIDDoc]] as described in the previous step 3.
+4. **Generate a preliminary DID Log Entry** JSON object containing the same JSON
+   items that will be in the published [[ref: DID log entry]], but with some
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:233]: Fixed

File: spec/specification.md [hunk 11]
```diff
@@ -215,74 +220,71 @@ Creating a `did:tdw` DID is done by carrying out the following steps.
...
+   values preset, pending calculation of the [[ref: SCID]] and [[ref entryHash]] and
+   without the `proof` item.
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:235]: Fixed

File: spec/specification.md [hunk 12]
```diff
@@ -215,74 +220,71 @@ Creating a `did:tdw` DID is done by carrying out the following steps.
...
+
+   1. The value of `versionId` string **MUST** be the placeholder literal `"{SCID}"`.
+   2. The value of `versionTime` string **MUST** be a valid [[ref: ISO8601]] date/time string,
+   and the represented time **MUST** be before or equal to the current time.
+   3. The value of the `parameters` item must be a JSON object defined at the
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:240]: Fixed

File: spec/specification.md [hunk 13]
```diff
@@ -215,74 +220,71 @@ Creating a `did:tdw` DID is done by carrying out the following steps.
...
+      discretion of the [[ref: DID Controller]]. The items in the object
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:241]: Fixed

File: spec/specification.md [hunk 14]
```diff
@@ -215,74 +220,71 @@ Creating a `did:tdw` DID is done by carrying out the following steps.
...
+      **MUST** be as permitted in the [DID Generation and Verification
+      Parameters](#didtdw-did-method-parameters) section of this specification,
+      and all required values in the first version of the DID **MUST** be
+      present. In addition, where the [[ref: SCID]] of the DID is referenced in
+      the parameters, the placeholder literal string `{SCID}` **MUST** be used
+      in place of the to-be-calculated [[ref: SCID]].
+   4. The value of the `state` item **MUST** be the initial [[ref: DIDDoc]] as
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:248]: Fixed

File: spec/specification.md [hunk 15]
```diff
@@ -215,74 +220,71 @@ Creating a `did:tdw` DID is done by carrying out the following steps.
...
-
-   6. **Add the [[ref: Data Integrity]] proof**
-   The fifth item in the input JSON array **MUST** be added to the JSON array. The result is in the
+   4. **Replace the preliminary `versionId` value** The value of the `versionId`
+   item **MUST** be updated with the literal string `1` (for version number 1),
+   a literal `-`, followed by the `entryHash` value calculated in the previous
+   step.
+   5. **Generate the [[ref: Data Integrity]] proof** A [[ref: Data Integrity]]
+   proof on the preliminary JSON object as updated in the previous step **MUST**
+   be generated using an authorized key in the required `updateKeys` item in the
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:272]: Fixed

File: spec/specification.md [hunk 16]
```diff
@@ -320,16 +322,14 @@ The following steps MUST be executed to resolve the [[ref: DIDDoc]] for a `did:t
 To process the retrieved [[ref: DID Log]] file, the resolver **MUST** carry out
 the following steps on each of the [[ref: log entries]] in the order they appear in the
 file, applying the [[ref: parameters]] set from the current and previous
-entries. As noted in the [DID Log File](#the-did-log-file) section, log entries
-are each a JSON array with five items:
+entries. As noted in the [DID Log File](#the-did-log-file) section, [[ref: log entries]]
+are each a JSON object with the following items:
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:326]: Fixed

File: spec/specification.md [hunk 17]
```diff
@@ -372,9 +372,8 @@ For each entry:
    (defined in the [[ref: parameters]]) according to the
    [SCID Generation and Verification](#scid-generation-and-verification) section
    of this specification.
-7. Extract the contents of the [[ref: DIDDoc]] for each entry by using the JSON value
-   of the `value` item, or by using [[ref: JSON Patch]] to apply the JSON
-   value of the `patch` entry item to the previous version of the [[ref: DIDDoc]].
+7. Get the value of the [[ref: log entry]] item `state`, which is the [[ref:
+   DIDDoc]] for the version.
 8. If [[ref: Key Pre-Rotation]] is being used, the hash of all `updateKeys` entries
    in the `parameters` item **MUST** match a hash in
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:378]: Fixed

File: spec/specification.md [hunk 18]
```diff
@@ -438,10 +437,10 @@ verifiable [[ref: DID Log Entry]] follows a similar process to the
       is forced to) publish the DID at a different Internet location and wants
       to retain the [[ref: SCID]] and history of the DID. For details, see the
       [DID Portability](#did-portability) section of this specification.
-2. Define a JSON array of [[ref: parameters]] that affect the evolution of the
+2. Define the [[ref: parameters]] JSON object to include the items that affect the evolution of the
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:440]: Fixed

File: spec/specification.md [hunk 19]
```diff
@@ -450,35 +449,28 @@ verifiable [[ref: DID Log Entry]] follows a similar process to the
      effect after the entry has been published. For example, rotating the keys
      authorized to update a DID or changing the [[ref: witnesses]] for a DID take effect
      only *after* the entry in which they are defined has been published.
-3. Generate a preliminary [[ref: DID Log Entry]] JSON array with the following JSON items:
-   1. The `versionId` **MUST** be `versionId` value from the previous [[ref: DID Log Entry]].
-   2. The `versionTime` value **MUST** be an [[ref: ISO8601]] format time. The time
-      **MUST** be greater than the time of the previous [[ref: log entry]], and **MUST** be less
-      than or equal to the current time.
+3. Generate a preliminary [[ref: DID log entry]] JSON object containing the following JSON items:
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:452]: Fixed

File: spec/specification.md [hunk 20]
```diff
@@ -450,35 +449,28 @@ verifiable [[ref: DID Log Entry]] follows a similar process to the
...
-      section of this specification.
-      1. An implementation **MAY** skip the [[ref: JSON Patch]] process and
-         simply put the full new version of the [[ref: DIDDoc]] in the item
-         `{"value": <DIDDoc>}` as is done in the initial entry in the log.
+   4. Set the `state` JSON object to be the new version of the [[ref: DIDDoc]].
 4. Calculate the new `versionId` of the new [[ref: DID Log Entry]], including
    incrementing the version number integer and using the process described in
    the [Entry Hash Generation and Verification](#entry-hash-generation-and-verification)
    section of this specification.
 5. Replace the value of the `versionId` item in the preliminary [[ref: DID Log
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:464]: Fixed

File: spec/specification.md [hunk 21]
```diff
@@ -450,35 +449,28 @@ verifiable [[ref: DID Log Entry]] follows a similar process to the
...
-   [Authorized Keys](#authorized-keys) section of this specification.
+6. Generate a [[ref: Data Integrity]] proof on the [[ref: DID log entry]] using
+   an authorized key, as defined in the [Authorized Keys](#authorized-keys)
+   section of this specification.
 7. If the [[ref: DID Controller]] has opted to use [[ref: witnesses]] for the
    DID, collect the required approvals from the DID's [[ref: witnesses]], adding
    their proofs to the [[ref: data integrity]] proof. See the [DID
    Witnesses](#did-witnesses) section of this specification.
-8. The proof array **MUST** be added as the fifth and last JSON item in the [[ref: log entry]].
+8. The proof JSON object **MUST** be added as the value of the `proof` item in the [[ref: log entry]].
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:473]: Fixed

File: spec/specification.md [hunk 22]
```diff
@@ -495,11 +487,12 @@ section of this specification.
 
 #### Deactivate (Revoke)
 
-To deactivate the DID, the [[ref: DID Controller]] **SHOULD** add to the [[ref:
+To deactivate the DID, the [[ref: DID Controller]] **MUST** add to the [[ref:
 DID log entry]] [[ref: parameters]] the item `"deactivated": true`. A [[ref: DID
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:491]: Fixed

File: spec/specification.md [hunk 23]
```diff
@@ -495,11 +487,12 @@ section of this specification.
...
-Controller]] **MAY** update the [[ref: DIDDoc]] further to indicate the deactivation of
-the DID, such as including an empty `updateKeys` list (`"updateKeys": []`) in
-the [[ref: parameters]], preventing further versions of the DID.
+Controller]] **SHOULD** update the [[ref: DIDDoc]] and `parameters` object to
+further indicate the deactivation of the DID, such as including an empty
+`updateKeys` list (`"updateKeys": []`) in the [[ref: parameters]], preventing
+further versions of the DID.
 
 A resolver encountering in the [[ref: DID log entry]] [[ref: parameters]] the
 item `"deactivated": true` **MUST** return in the [[ref: DIDDoc]] Metadata the JSON item
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:498]: Fixed

File: spec/specification.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:683]: Fixed

File: spec/specification.md [hunk 24]
```diff
@@ -677,13 +674,13 @@ To verify the [[ref: SCID]] of a `did:tdw` DID being resolved, the resolver
...
-2. Extract the `scid` item value from the [[ref: parameters]] in the [[ref: DID log entry]]. 
+2. Extract the `scid` item value from the [[ref: parameters]] in the [[ref: DID log entry]].
 3. Determine the hash algorithm used by the [[ref: DID Controller]] from the [[ref: multihash]] `scid` value.
    - The hash algorithm **MUST** be one listed in the
    [parameters](#didtdw-did-method-parameters) defined by the version of the
    `did:tdw` specification being used by the [[ref: DID Controller]] based on the
    `method` [[ref: parameters]] item.
-4. Remove the [[ref: data integrity]] proof item from the first [[ref: DID log entry]].
+4. Remove the [[ref: data integrity]] proof item from the [[ref: DID log entry]].
 5. Replace the `versionId` item's value with the literal `"{SCID}"`.
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:684]: Fixed

File: spec/specification.md [hunk 25]
```diff
@@ -727,28 +724,28 @@ produce an [[ref: entry hash]]. As this is a first entry in a [[ref: DID Log]],
 `entryHash` (first item) is the [[ref: SCID]] of the DID.
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:724]: Fixed

File: spec/specification.md [hunk 26]
```diff
@@ -727,28 +724,28 @@ produce an [[ref: entry hash]]. As this is a first entry in a [[ref: DID Log]],
...
-1. Extract the `versionId` (first item) in the [[ref: DID log entry]], and
+1. Extract the `versionId` in the [[ref: DID log entry]], and
    remove from it the version number and dash prefix, leaving the log entry
-   `entryHash`.
+   `entryHash` value.
 2. Determine the hash algorithm used by the [[ref: DID Controller]] from the [[ref: multihash]] `entryHash` value.
    - The hash algorithm **MUST** be one listed in the
    [parameters](#didtdw-did-method-parameters) defined by the version of the
    `did:tdw` specification being used by the [[ref: DID Controller]] based on the
    `method` [[ref: parameters]] item set in the current or most recent prior [[ref: log entry]].
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:744]: Fixed

File: spec/specification.md [hunk 27]
```diff
@@ -947,9 +908,8 @@ An overview of the [[ref: witness]] mechanism is as follows:
     deployment of `did:tdw`. That is up to the ecosystem in which the [[ref: DID
     Controller]] and [[ref: witnesses]] are participating.
 - If the verification is successful and the approval granted, the [[ref:
-  witness]] sends a [[ref: Data Integrity]] proof to the [[ref: DID Controller]] similar
-  to that generated by the [[ref: DID Controller]], with the [[ref: DIDDoc]] as the signed object,
-  the entry log's `versionId` as the challenge, but signed by the [[ref: witnesses]] key.
+  witness]] sends a [[ref: Data Integrity]] proof across the [[ref: DID log entry]] to the [[ref: DID Controller]], similar
+  to that generated by the [[ref: DID Controller]], but signed by the [[ref: witness]]'s key.
 - When a weighted threshold of proofs are received, the DID Controller
   inserts the [[ref: witnesses]]'s proofs into the array of proofs that are the last item in
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/specification.md:914]: Fixed — adjusted slightly.

**Comment by swcurran** [comment]
Applied all the changes noted by @martipos and then removed the remaining instances of “item” used in the specification and related files.  I had thought that was the right term in JSON.  I’ve been using it for a while in many places, so I have now retrained myself! :-)

**Comment by swcurran** [comment]
Added to the specification that the `versionTime` timestamp MUST be UTC time zone — per the suggestion in #112.

**Review by brianorwhatever — APPROVED** [review]

File: spec/overview.md [hunk 28]
```diff
@@ -61,31 +61,29 @@ The following is a `tl;dr` summary of how `did:tdw` works:
...
-    1. The `versionId` of the entry, a value that combines the version number
-       (starting at `1` and incrementing by one per version), a literal dash
+3. Each [[ref: DID log entry]] is a JSON object containing the following properties:
+    1. `versionId` -- a value that combines the version number
+       (starting at `1` and incremented by one per version), a literal dash
        `-`, and a hash of the entry. The [[ref: entry hash]] calculation links each entry
        to its predecessor in a ledger-like chain.
-    2. The `versionTime` (as stated by the [[ref: DID Controller]]) of the entry.
+    2. `versionTime` -- as asserted by the [[ref: DID Controller]].
     3. A set of `parameters` that impact the processing of the current and
```
Comment (PatStLouis) [inline spec/overview.md:70]:
```suggestion
    3. `parameters` -- a set of parameters that impact the processing of the current and
```

File: spec/overview.md [hunk 29]
```diff
@@ -61,31 +61,29 @@ The following is a `tl;dr` summary of how `did:tdw` works:
...
-4. In generating the first version of the [[ref: DIDDoc]], the [[ref: DID Controller]] calculates
-  the [[ref: SCID]] for the DID from the entire first [[ref: log entry]] (which
-  includes the [[ref: DIDDoc]]) by adding the string {SCID} everywhere the actual [[ref: SCID]]
-  is to be placed. The [[ref: DID Controller]] then replaces these placeholders
-  with the just calculated [[ref: SCID]], including it as a `parameter` in the first [[ref: log
-  entry]], and inserting it where needed in the initial (and all subsequent)
-  DIDDocs. The [[ref: SCID]] enables an optional [[ref: portability]] capability, allowing a DID's
-  web location to be moved to a new location while retaining the DID and version
-  history of the DID.
+    4. The `state`, the new version of the [[ref: DIDDoc]].
```
Comment (PatStLouis) [inline spec/overview.md:74]:
```suggestion
    4. `state` -- the new version of the [[ref: DIDDoc]].
```

File: spec/overview.md [hunk 30]
```diff
@@ -61,31 +61,29 @@ The following is a `tl;dr` summary of how `did:tdw` works:
...
+    5. A [[ref: Data Integrity]] (DI) proof across the entry, signed by a [[ref:
```
Comment (PatStLouis) [inline spec/overview.md:75]:
```suggestion
    5. `proof` -- a [[ref: Data Integrity]] (DI) proof across the entry, signed by a [[ref:
```

**Review by PatStLouis — CHANGES_REQUESTED** [review]
Overall great readability improvements and update. Added a few change request for consistency.

**Review by PatStLouis — APPROVED** [review]

**Pull request merged**